source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp

```diff
@@ -128,180 +127,119 @@ MPxNode::SchedulingType MaterialXNode::schedulingType() const
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Vector3>::TYPE)
-				{
-                    float3& value = dataHandle.asFloat3();
-                    valueElement->setValue(MaterialX::Vector3(value[0], value[1], value[2]));
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Vector4>::TYPE)
-				{
-                    MFloatVector& value = dataHandle.asFloatVector();
-                    valueElement->setValue(MaterialX::Vector4(value[0], value[1], value[2], value[3]));
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Color2>::TYPE)
-				{
-                    float2& value = dataHandle.asFloat2();
-                    valueElement->setValue(MaterialX::Color2(value[0], value[1]));
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Color3>::TYPE)
-				{
-					float3& value = dataHandle.asFloat3();
-					valueElement->setValue(MaterialX::Color3(value[0], value[1], value[2]));
-				}
-				else if (type == MaterialX::TypedValue<MaterialX::Color4>::TYPE)
-				{
-					MFloatVector& value = dataHandle.asFloatVector();
-                    valueElement->setValue(MaterialX::Color4(value[0], value[1], value[2], value[3]));
-				}
-				else if (type == MaterialX::TypedValue<float>::TYPE)
-				{
-					float& value = dataHandle.asFloat();
-					valueElement->setValue(value);
-				}
-			}
-		}
-	}
-	return true;
+    return true;
 }
 
-void MaterialXNode::setAttributeValue(MObject &materialXObject, MObject &attr, float* values, unsigned int size, MDGModifier& mdgModifier)
+bool MaterialXNode::setInternalValue(const MPlug& plug, const MDataHandle& dataHandle)
 {
-	MPlug plug(materialXObject, attr);
-	if (size == 1)
-	{
-		mdgModifier.newPlugValueDouble(plug, values[0]);
-	}
-	else
-	{
-		for(unsigned int i=0; i<size; i++)
-		{
-			MPlug indexPlug = plug.child(i);
-			mdgModifier.newPlugValueDouble(indexPlug, values[i]);
-		}
-	}
-}
+    auto displayError = [this](const char* error)
+    {
+        MString message("Failed to create shader for ");
+        message += typeName();
+        message += ' ';
+        message += name();
+        message += ": ";
+        message += error;
+        MGlobal::displayError(message);
+    };
+
+    auto loadDocument = [this]()
+    {
+        return MaterialXMaya::loadDocument(
+            _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
+        );
+    };
 
-void MaterialXNode::createAttributesFromDocument(MDGModifier& mdgModifier)
-{
-    MaterialX::DocumentPtr document;
-    if (!materialXData || !(document = materialXData->getDocument()))
+    if (plug == DOCUMENT_ATTRIBUTE)
+    {
+        const MString& value = dataHandle.asString();
+        if (_documentFilePath == value)
+        {
+            return true;
+        }
+
+        _documentFilePath = value;
+        _materialXData.reset();
+
+        try
+        {
+            _materialXData.reset( new MaterialXData(
+                loadDocument(),
+                _xmlElementPath.asChar(),
```

Comment:
It's debatable if the element path should be reused. It's a reference to the last document. I would say we only reuse this if we are setting to the same document name again (this could be a "reload" case). Otherwise we clear this and find the first renderable element or just use a fallback shader. I think finding the first renderable element is "nicer" to the user.

Reply:
BTW since we always enumerate renderable elements when editing either the document or the element, it seems easy to implement a drop-down list for the element attribute, just like the technique attribute in the glslShader plugin.